src/charm.py

```diff
@@ -206,34 +205,34 @@ def _on_collect_unit_status(self, event: ops.CollectStatusEvent) -> None:
         """Handle collect unit status event (called after every event)."""
         if not self.model.relations.get("credentials"):
             event.add_status(BlockedStatus("Keystone is not related"))
+            return
 
         if not self._get_keystone_data():
             event.add_status(WaitingStatus("Waiting for credentials from keystone"))
+            return
 
         if not self.model.relations.get("cos-agent"):
             event.add_status(BlockedStatus("Grafana Agent is not related"))
-
-        upstream_snap = get_installed_snap_service(UPSTREAM_SNAP)
+            return
 
         # this is necessary when doing a charm upgrade coming from revision 27
+        upstream_snap = get_installed_snap_service(UPSTREAM_SNAP)
         if upstream_snap.present:
             event.add_status(
                 BlockedStatus(
                     "golang-openstack-exporter detected. Please see: "
                     "https://charmhub.io/openstack-exporter#known-issues"
                 )
             )
+            return
 
         snap_service = get_installed_snap_service(SNAP_NAME)
 
         if not snap_service.present:
-            event.add_status(
-                BlockedStatus("snap service is not installed, please check snap service")
-            )
-        elif not snap_service.is_active():
-            event.add_status(
-                BlockedStatus("snap service is not running, please check snap service")
-            )
+            raise RuntimeError(f"{SNAP_NAME} snap is not installed, but it should be")
```

Comment:
Will we see this message when we type `juju status`?

Reply:
Unfortunately not - it's a limitation of juju that it's not possible to set a juju status for error state.  It will look something like this:

```
openstack-exporter/0*  error     idle   0        10.168.106.56             hook failed: "update-status"
```

Reply:
I think I'm going back and forth on blocked vs error status here... I'm not sure what is best. :/

Reply:
I have two assumptions:

1. The configuration event handler won’t take effect until the installation is successful.
2. We can provide guidance to the user based on the most common scenario

If we've already handled the installation procedure in snap_install_or_refresh(), why do we need to check these two cases here? Is it possible that this handler could be triggered before the installation is complete? If so, would it be reasonable to provide general guidance to help users resolve the issue?

For example: 'Snap service is unavailable. Please wait for the installation to complete or manually reinstall the Smartctl snap if the issue persists.'

Reply:
> If we've already handled the installation procedure in snap_install_or_refresh(), why do we need to check these two cases here?

I guess it's to check if something has broken the exporter - like maybe the exporter crashed, or a colocated application/charm interfered, etc.

I like the idea of a more helpful message in juju status.

Reply:
I guess the point of concern is that if the snap service stops running or the snap is uninstalled, it's likely to be a deeper issue that couldn't be resolved simply by reinstalling or restarting manually. :thinking:

Reply:
Updated